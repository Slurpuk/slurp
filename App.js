import 'react-native-gesture-handler';
import React, {useEffect, useState} from 'react';
import {NavigationContainer} from '@react-navigation/native';
import HamburgerSlideBarNavigator from './src/navigation/HamburgerSlideBarNavigator';
import SignUpPage from './src/screens/SignUpPage';
import LogInPage from './src/screens/LogInPage';
import WelcomePages from './src/screens/WelcomePages';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import AsyncStorage from '@react-native-async-storage/async-storage';
import firebase from '@react-native-firebase/app';
import firestore from '@react-native-firebase/firestore';
import {Alert} from 'react-native';
import auth from '@react-native-firebase/auth';

export const GlobalContext = React.createContext();
export default function App() {
  const [isFirstTime, setIsFirstTime] = useState(true);
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [currentUser, setCurrentUser] = useState(auth().currentUser);
  const [userRef, setUserRef] = useState(null);
  const [userObj, setUserObj] = useState(null);
  const [shopsData, setShopsData] = useState([]);
  const [isShopIntro, setIsShopIntro] = useState(false);
  const [currShop, setCurrShop] = useState(shopsData[0]);
  const [isFullScreen, setFullScreen] = useState(false);
  const [basketContent, setBasketContent] = useState([]);
  const [basketSize, setBasketSize] = useState(0);
  const [total, setTotal] = useState(0);

  const checkForFirstTime = async () => {
    const result = await AsyncStorage.getItem('isFirstTime');
    //if what we get from the Async is null we are opening the app for the first time
    //if we pressed the sign up button on the last slide we set the 'isFirstTime' to 'no'
    if (result === null) {
      setIsFirstTime(true);
    } //now we can use the isFirstTimeLoad state to choose what to render
  };

  useEffect(() => {
    checkForFirstTime();
  }, []);

  useEffect(() => {
    const subscriber = firebase.auth().onAuthStateChanged(user => {
      if (user) {
        setIsLoggedIn(true);
        setCurrentUser(user);
        setUser();
      } else {
        setIsLoggedIn(false);
        setCurrentUser(null);
      }
    });
    // Unsubscribe from events when no longer in use
    return () => subscriber();
  });

  const enterApp = () => {
    setIsFirstTime(false);
    AsyncStorage.setItem('isFirstTime', 'potatoesInPower');
  };

  function clearBasket() {
    setBasketContent([]);
    setBasketSize(0);
    setTotal(0);
  }

  function newShop({shop, navigation}) {
    clearBasket();
    setCurrShop(shop);
    navigation.navigate('Shop page');
  }

  useEffect(() => {
    const subscriber = firestore()
      .collection('Users')
      .doc(userRef)
      .onSnapshot(documentSnapshot => {
        setUserObj(documentSnapshot.data());
      });

    // Stop listening for updates when no longer required
    return () => subscriber();
  }, [userRef]);

  async function setUser() {
    if (currentUser) {
      await firestore()
        .collection('Users')
        .where('authID', '==', currentUser.uid)
        .get()
        .then(querySnapshot => {
          querySnapshot.forEach(documentSnapshot => {
            setUserRef(documentSnapshot.id);
          });
        });
    }
  }

  function changeShop({shop, navigation}) {
    if (currShop !== shop && basketSize !== 0) {
      Alert.alert(
        'Are you sure ?',
        'Changing shops will clear your basket.',
        [
          {
            text: 'Yes',
            onPress: () => newShop({shop, navigation}),
          },
          {
            text: 'No',
            onPress: () => navigation.navigate('Shop page'),
            style: 'cancel',
          },
        ],
        {cancelable: false},
      );
    } else {
      setCurrShop(shop);
      navigation.navigate('Shop page');
    }
  }

  // Subscribe to the Shops model
  useEffect(() => {
    const subscriber = firestore()
      .collection('CoffeeShop')
      .onSnapshot(querySnapshot => {
        const shops = [];

        querySnapshot.forEach(documentSnapshot => {
          let shopData = {
            ...documentSnapshot.data(),
            key: documentSnapshot.id,
          };

          let coffees = [];
          let drinks = [];
          let snacks = [];
          documentSnapshot.data().ItemsOffered.forEach(itemRef => {
            firestore()
              .doc(itemRef.path)
              .onSnapshot(querySnapshot => {
                let collection = '';
                if (itemRef.path.includes('Coffees')) {
                  collection = coffees;
                } else if (itemRef.path.includes('Drinks')) {
                  collection = drinks;
                } else {
                  collection = snacks;
                }
                collection.push({
                  ...querySnapshot.data(),
                  key: querySnapshot.id,
                });
              });
          });
          shopData.ItemsOffered = {
            Coffees: coffees,
            Drinks: drinks,
            Snacks: snacks,
          };
          shops.push(shopData);
          setShopsData(shops);
          setCurrShop(shops[0]);
        });
      });

    // Unsubscribe from events when no longer in use
    return () => subscriber();
  }, []);

  function isSameItem(it, currIt) {
    if (currIt.hasOwnProperty('Bean') && it.hasOwnProperty('Bean')) {
      let itOptions = '';
      it.options.forEach(option => (itOptions += option.Name));
      let currItOptions = '';
      currIt.options.forEach(option => (currItOptions += option.Name));
      return it.key === currIt.key && itOptions === currItOptions;
    } else {
      return it.key === currIt.key;
    }
  }

  function addToBasket(item) {
    const basket = basketContent;
    const exist = basket.find(x => isSameItem(x, item));
    if (exist) {
      setBasketContent(
        basket.map(x =>
          isSameItem(x, item) ? {...exist, count: exist.count + 1} : x,
        ),
      );
    } else {
      setBasketContent([...basket, {...item, count: 1}]);
    }
    setTotal(total + item.Price);
    setBasketSize(basketSize + 1);
  }

  function removeFromBasket(item) {
    const basket = basketContent;
    const exist = basket.find(x => isSameItem(x, item));
    if (exist.count === 1) {
      setBasketContent(basket.filter(x => x.key !== item.key));
    } else {
      setBasketContent(
        basket.map(x =>
          isSameItem(x, item) ? {...exist, count: exist.count - 1} : x,
        ),
      );
    }
    setTotal(total - item.Price);
    setBasketSize(basketSize - 1);
  }

  const setShopIntro = () => {
    setIsShopIntro(!isShopIntro);
  };

  const Stack = createNativeStackNavigator();
  return (
    <GlobalContext.Provider
      value={{
        enterApp: enterApp,
        user: currentUser, // Returns the authentication object
        currShop: currShop,
        setCurrShop: changeShop,
        isShopIntro: isShopIntro,
        shopsData: shopsData,
        setShopIntro: setShopIntro,
        isFullScreen: isFullScreen,
        setFullScreen: setFullScreen,
        basketContent: basketContent,
        setBasketContent: setBasketContent,
        total: total,
        setTotal: setTotal,
        addToBasket: addToBasket,
        removeFromBasket: removeFromBasket,
        basketSize: basketSize,
        clearBasket: clearBasket,
        currentUser: userObj, // Returns the model object
        userRef: userRef,
      }}
    >
      <NavigationContainer>
        {isLoggedIn ? (
          <HamburgerSlideBarNavigator />
        ) : (
          <Stack.Navigator
            screenOptions={{
              headerShown: false,
            }}
          >
            {isFirstTime ? (
              <Stack.Screen name="Welcome" component={WelcomePages} />
            ) : null}
            <Stack.Screen name="LogIn" component={LogInPage} />
            <Stack.Screen name="SignUp" component={SignUpPage} />
          </Stack.Navigator>
        )}
      </NavigationContainer>
    </GlobalContext.Provider>
  );
}
