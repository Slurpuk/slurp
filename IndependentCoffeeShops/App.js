import React, {useState} from 'react';
import {Pressable} from 'react-native';
import {BottomSheet} from 'react-native-bottomsheet-reanimated';
import type {Node} from 'react';
//import Inputs from 'DesignMap/src/screens/inputs.js'
import {
  View,
  Text,
  TouchableOpacity,
  TextInput,
  StyleSheet,
} from 'react-native';
import PrimaryButton from './PrimaryButton';
import textStyles from './stylesheets/textStyles';
import {authentication} from './firestore/firebase-config';
import firebase from '@react-native-firebase/app';
import auth from '@react-native-firebase/auth';
import firestore from '@react-native-firebase/firestore';
import SignUpPage from "./components/SignUpPage";
import LoginPage from "./components/LoginPage";
import {rgbaColor} from 'react-native-reanimated/src/reanimated2/Colors';

export default function App() {
  const usersCollection = firestore().collection('Users');
  const [isSignedIn, setIsSignedIn] = useState(false);
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [x, setX] = useState(usersCollection);

  const registerUser = () => {
    auth()
      .createUserWithEmailAndPassword(email, password)
      .then(re => {
        console.log(re);
        console.log(x);
      })
      .catch(re => {
        console.log(re);
      })
  }

  return (
    <View style={styles.mainContainer}>
      <TextInput
        placeholder="email"
        value={email}
        onChangeText={text => setEmail(text)}
      />
      <TextInput
        placeholder="password"
        value={password}
        secureTextEntry={true}
        onChangeText={text => setPassword(text)}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  mainContainer: {
    marginTop: 55,
    flex: 1,
    padding: 10,
  },
});
