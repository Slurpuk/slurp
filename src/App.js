import 'react-native-gesture-handler';
import React, {useContext, useEffect, useReducer} from 'react';
import {NavigationContainer} from '@react-navigation/native';
import HamburgerSlideBarNavigator from './navigation/HamburgerSlideBarNavigator';
import SignUpPage from './screens/SignUpPage';
import LogInPage from './screens/LogInPage';
import WelcomePages from './screens/WelcomePages';
import {createNativeStackNavigator} from '@react-navigation/native-stack';
import firestore from '@react-native-firebase/firestore';
import auth from '@react-native-firebase/auth';
import LoadingPage from './screens/LoadingPage';
import {Alerts} from './data/Alerts';
import {
  formatCurrentOrders,
  formatOrders,
  formatPastOrders,
  getFormattedShops,
  loginUser,
  logoutUser,
  refreshOrders,
  refreshCurrentShop,
} from './helpers/screenHelpers';
import {getIsFirstTime, refreshCurrentBasket} from './helpers/storageHelpers';
import {firebase} from '@react-native-firebase/database';
import {GlobalAction} from './data/actionEnum';
import {GlobalContext} from './contexts';
import {globalReducer} from './reducers';
import {CurrentOrderStatuses, PastOrderStatuses} from './data/OrderStatus';

/**
 * Root component rendered when the application boots.
 */
export default function App() {
  const initialState = useContext(GlobalContext);
  const [globalState, globalDispatch] = useReducer(globalReducer, initialState);
  const LoggedOutStack = createNativeStackNavigator(); // Stack navigator for logged out users.
  const DB_URL =
    'https://slurp-59784-default-rtdb.europe-west1.firebasedatabase.app/';

  /**
   * Side effect that fires when App first renders to determine whether it is the first time the app is used since download
   * Sets the first time state accordingly.
   */
  useEffect(() => {
    async function setIsFirstTime() {
      const isFirstTime = await getIsFirstTime();
      if (!isFirstTime) {
        globalDispatch({type: GlobalAction.USE_APP});
      }
    }

    setIsFirstTime().catch(() => Alerts.elseAlert());
  }, []);

  /**
   * Side effect that tracks the authentication state of the current user.
   * Sets the current user object accordingly.
   */
  useEffect(() => {
    const subscriber = auth().onAuthStateChanged(async user => {
      if (user) {
        await loginUser(user, globalDispatch);
      } else {
        await logoutUser(globalDispatch);
      }
    });
    // Unsubscribe from events when no longer in use
    return () => subscriber();
  }, []);

  /**
   * Side effect that tracks changes in the model instance of the user in the database and updates the state accordingly
   * Sets the current user object accordingly.
   */
  useEffect(() => {
    if (!globalState.isLoadingUser) {
      const subscriber = firestore()
        .collection('users')
        .where('email', '==', auth().currentUser.email)
        .onSnapshot(query => {
          let newUserDoc = query.docs[0];
          let newUser = newUserDoc.data();
          globalDispatch({
            type: GlobalAction.SET_CURRENT_USER,
            user: {...newUser, key: newUserDoc.id, ref: newUserDoc.ref},
          });
        });
      return () => subscriber();
    }
  }, [globalState.isLoadingUser]);

  /**
   * Side effect that tracks changes in the current orders of the logged-in user.
   */
  useEffect(() => {
    if (!globalState.isLoadingUser) {
      const subscriber = firestore()
        .collection('orders')
        .where('is_displayed', '==', true)
        .where('user', '==', globalState.currentUserRef)
        .where('status', 'in', CurrentOrderStatuses)
        .onSnapshot(async query => {
          let currentOrders = await formatCurrentOrders(
            formatOrders(query.docs),
          );
          globalDispatch({
            type: GlobalAction.SET_CURRENT_ORDERS,
            orders: currentOrders,
          });
        });
      return () => subscriber();
    }
  }, [globalState.currentUserRef, globalState.isLoadingUser]);

  /**
   * Side effect that tracks changes in the past orders of the logged-in user.
   */
  useEffect(() => {
    if (!globalState.isLoadingUser) {
      const subscriber = firestore()
        .collection('orders')
        .where('is_displayed', '==', true)
        .where('user', '==', globalState.currentUserRef)
        .where('status', 'in', PastOrderStatuses)
        .onSnapshot(async query => {
          let pastOrders = await formatPastOrders(formatOrders(query.docs));
          globalDispatch({
            type: GlobalAction.SET_PAST_ORDERS,
            orders: pastOrders,
          });
        });
      return () => subscriber();
    }
  }, [globalState.currentUserRef, globalState.isLoadingUser]);

  /**
   * Side effect that tracks any change in the coffee shop model.
   * Retrieves the latest shop data, formats it and updates the current set of shops.
   */
  useEffect(() => {
    if (!globalState.isLoadingUser) {
      const subscriber = firestore()
        .collection('coffee_shops')
        .onSnapshot(async querySnapshot => {
          let formattedShops = await getFormattedShops(querySnapshot);
          globalDispatch({
            type: GlobalAction.SET_LIST_OF_SHOPS,
            shops: formattedShops,
          });
          await refreshCurrentShop(
            formattedShops,
            globalState.currentShopKey,
            globalDispatch,
          );
        });
      // Unsubscribe from events when no longer in use
      return () => subscriber();
    }
  }, [globalState.currentShopKey, globalState.isLoadingUser]);

  /**
   * Side effect that refreshes the logged-in user's data once the App loads.
   */
  useEffect(() => {
    async function refreshUserData() {
      if (!globalState.isLoadingUser) {
        await refreshCurrentBasket(globalDispatch);
        await refreshOrders(globalState.currentUserRef, globalDispatch);
      }
    }

    refreshUserData().then(() => console.log('refreshed'));
  }, [globalState.currentUserRef, globalState.isLoadingUser]);

  /**
   * Side effect that tracks the connection state of the app.
   */
  useEffect(() => {
    if (!globalState.isLoadingUser) {
      const uid = firebase.auth().currentUser.uid;

      const userStatusFirestoreRef = firestore()
        .collection('connection_status')
        .doc(uid);

      const userStatusDatabaseRef = firebase
        .app()
        .database(DB_URL)
        .ref('/status/' + uid);

      const isOfflineForDatabase = {
        status: 'offline',
        last_changed: firebase.database.ServerValue.TIMESTAMP,
      };
      const isOnlineForDatabase = {
        status: 'online',
        last_changed: firebase.database.ServerValue.TIMESTAMP,
      };
      const isOfflineForFirestore = {
        status: 'offline',
        last_changed: firebase.firestore.FieldValue.serverTimestamp(),
      };
      const isOnlineForFirestore = {
        status: 'online',
        last_changed: firebase.firestore.FieldValue.serverTimestamp(),
      };

      firebase
        .app()
        .database(DB_URL)
        .ref('.info/connected')
        .on('value', function (snapshot) {
          if (snapshot.val() === false) {
            userStatusFirestoreRef.set(isOfflineForFirestore);
            return;
          }
          userStatusDatabaseRef
            .onDisconnect()
            .set(isOfflineForDatabase)
            .then(async function () {
              userStatusDatabaseRef.set(isOnlineForDatabase);
              await userStatusFirestoreRef.set(isOnlineForFirestore);
            });
        });

      const subscriber = userStatusFirestoreRef.onSnapshot(function (doc) {
        const isOnline = doc.data().status === 'online';
        isOnline
          ? globalDispatch({type: GlobalAction.CONNECT})
          : globalDispatch({type: GlobalAction.DISCONNECT});
      });

      return () => subscriber();
    }
  }, [globalState.isLoadingUser]);

  return (
    <GlobalContext.Provider value={{globalState, globalDispatch}}>
      <NavigationContainer>
        {!globalState.isLoadingUser ? (
          <HamburgerSlideBarNavigator />
        ) : auth().currentUser ? (
          <LoadingPage />
        ) : (
          <LoggedOutStack.Navigator
            screenOptions={{
              headerShown: false,
            }}>
            {!globalState.isFirstTime ? (
              <LoggedOutStack.Screen name="Welcome" component={WelcomePages} />
            ) : null}
            <LoggedOutStack.Screen
              name="LogIn"
              children={props => <LogInPage {...props} />}
            />
            <LoggedOutStack.Screen
              name="SignUp"
              children={props => <SignUpPage {...props} />}
            />
          </LoggedOutStack.Navigator>
        )}
      </NavigationContainer>
    </GlobalContext.Provider>
  );
}
