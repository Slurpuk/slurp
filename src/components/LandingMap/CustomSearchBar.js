import React, {useCallback, useContext, useEffect, useState} from 'react';
import {SearchBar} from 'react-native-elements';
import {
  View,
  Text,
  StyleSheet,
  FlatList,
  Platform,
  Dimensions,
  Pressable,
  Keyboard,
} from 'react-native';

import {GlobalContext} from '../../../App';
import textStyles from '../../../stylesheets/textStyles';
const screenHeight = Dimensions.get('window').height;

const CustomSearchBar = ({searchBarFocused, setSearchBarFocussed}) => {
  const context = useContext(GlobalContext);

  //Used to store current text in search bar
  const [query, setQuery] = useState('');

  //the filtered list of shops to be shown in search results
  const [shops, setShops] = useState([]);

  //focus the search bar when a letter is typed and update the filtered shops
  const updateQuery = input => {
    if (!searchBarFocused && input !== '') {
      setSearchBarFocussed(true);
    } else if (searchBarFocused && input === '') {
      setSearchBarFocussed(false);
    }
    setQuery(input);
    filterShops();
  };

  //set the shops to be displayed to those which match the search query
  const filterShops = useCallback(() => {
    setShops(
      context.shopsData.filter(shop =>
        shop.name.toLowerCase().includes(query.toLowerCase()),
      ),
    );
  }, [context.shopsData, query]);

  //required to force update of search results to match current state, not previous state.
  useEffect(() => {
    filterShops();
  }, [filterShops, query]);

  //unfocus the search bar (remove search results)
  const clear = () => {
    setSearchBarFocussed(false);
  };

  //set the bottom sheet to the selected shop
  const selectShop = async shop => {
    if (shop.is_open) {
      await context.changeShop(shop);
      setSearchBarFocussed(false);
    }
  };

  return (
    <View style={styles.view}>
      <SearchBar
        placeholder="Type in a coffee shop..."
        // Function to run when a letter is pressed
        onChangeText={updateQuery}
        value={query}
        containerStyle={styles.container}
        inputContainerStyle={styles.inputContainer}
        // function to run when the x in search bar is pressed
        onClear={clear}
        placeholderTextColor={'#046D66'}
        inputStyle={{color: '#046D66'}}
        leftIconContainerStyle={{color: '#046D66'}}
        rightIconContainerStyle={{color: '#046D66'}}
        searchIcon={false}
        testID={'search_bar'}
      />
      {/*Only display search results when the search bar is focused*/}
      {searchBarFocused ? (
        <View style={styles.activeElementsWrapper}>
          <View style={styles.cover} />
          {shops.length > 0 ? (
            <FlatList
              // Only display the filtered shops in the results
              keyboardShouldPersistTaps="handled"
              data={shops}
              styles={styles.flatList}
              renderItem={({item}) => {
                return (
                  <View
                    style={[
                      styles.searchResultContainer,
                      {display: searchBarFocused ? 'flex' : 'none'},
                    ]}
                  >
                    <Pressable
                      onPressIn={() => {
                        if (item.is_open) {
                          Keyboard.dismiss();
                        }
                      }}
                      onPress={() => selectShop(item)}
                      style={({pressed}) => [
                        {
                          backgroundColor:
                            pressed && item.is_open
                              ? '#087562'
                              : item.is_open
                              ? 'white'
                              : '#C5C5C5',
                        },
                        styles.searchResult,
                      ]}
                      testID={'search_item_' + item.name}
                    >
                      {({pressed}) => (
                        <Text
                          style={[
                            {
                              color:
                                pressed && item.is_open ? 'white' : 'black',
                            },

                            styles.flatListItem,
                          ]}
                        >
                          {item.name}
                        </Text>
                      )}
                    </Pressable>
                  </View>
                );
              }}
            />
          ) : (
            <Text style={[textStyles.greyPoppins, styles.noResultsText]}>
              Sorry! {'\n\n'}
              No shop goes by this name. {'\n'}
              Please try a different one.
            </Text>
          )}
        </View>
      ) : null}
    </View>
  );
};
const screenWidth = Dimensions.get('window').width;

const styles = StyleSheet.create({
  view: {
    position: 'relative',
    marginRight: 0,
    ...Platform.select({
      ios: {
        marginTop: '0%',
      },
      android: {
        marginTop: '0%',
      },
    }),
    alignSelf: 'flex-end',
    flex: 1,
  },

  searchResultContainer: {
    width: screenWidth * 0.7,
  },

  container: {
    borderRadius: 20,
    borderColor: '#046D66',
    borderWidth: 2,
    borderBottomWidth: 2,
    borderTopWidth: 2,
    borderBottomColor: '#046D66',
    borderTopColor: '#046D66',
    backgroundColor: 'whitesmoke',
    height: 0.07 * screenHeight,
    color: '#046D66',
    overflow: 'visible',
    marginLeft: '3%',
    marginRight: '3%',
    zIndex: 0,
  },

  activeElementsWrapper: {
    marginLeft: -screenHeight * 0.1,
    maxHeight: screenHeight * 0.52,
  },

  searchResult: {
    width: screenWidth * 1.1,
  },

  cover: {
    backgroundColor: 'white',
    opacity: 0.8,
    width: screenWidth * 1.1,
    height: screenHeight * 0.15,
    position: 'absolute',
    top: -screenHeight * 0.15,
    zIndex: -1,
  },

  inputContainer: {
    borderRadius: 20,
    backgroundColor: 'whitesmoke',
    height: 0.05 * screenHeight,
    color: '#046D66',
  },
  flatListItem: {
    width: screenWidth * 1,
    maxWidth: screenWidth * 1,
    fontSize: 16,
    textAlignVertical: 'center',
    textAlign: 'center',
    paddingVertical: 15,
    borderRadius: 0,
    fontFamily: 'Poppins',
    borderBottomColor: '#DDDDDD',
    borderBottomWidth: 1,
    position: 'relative',
    height: 54,
  },
  flatList: {
    display: 'flex',
    overflow: 'visible',
  },
});

export default CustomSearchBar;
