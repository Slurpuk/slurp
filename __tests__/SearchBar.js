import {render, fireEvent} from '@testing-library/react-native';
import React from 'react';
import {GlobalContext} from '../App';
import CustomSearchBar from "../src/components/LandingMap/CustomSearchBar";
import {getByTestId, getByText, queryAllByText} from "@testing-library/react";
import customSearchBar from "../src/components/LandingMap/CustomSearchBar";
// import {fireEvent} from "@testing-library/react";

describe('search bar', function () {

  const globalContextMock = {
    shopsData: [{Name: 'Yucky Starbucks', IsOpen: true}, {Name: 'Black Goat', IsOpen: true}],
    switchShop: jest.fn(),
  };

  it('should render the search bar', function () {

    const {toJSON} = render(
      <GlobalContext.Provider value={globalContextMock}>
        <CustomSearchBar searchBarFocused={true} setSearchBarFocussed={jest.fn()}/>
      </GlobalContext.Provider>,
    );
    expect(true).toBe(true);
    expect(toJSON()).toMatchSnapshot();
  });

  it('Should be able to click on the search bar to focus it', function (){

    const {getByTestId, getByText, queryAllByText, rerender} = render(
        <GlobalContext.Provider value={globalContextMock}>
          <CustomSearchBar searchBarFocused={false} setSearchBarFocussed={jest.fn()}/>
        </GlobalContext.Provider>,
    );

    expect(queryAllByText('Yucky Starbucks')).toHaveLength(0);
    fireEvent(getByTestId('search-bar'), 'press');

    rerender(
        <GlobalContext.Provider value={globalContextMock}>
          <CustomSearchBar searchBarFocused={true} setSearchBarFocussed={jest.fn()}/>
        </GlobalContext.Provider>
    );

    expect(getByText('Yucky Starbucks')).toBeTruthy();

    fireEvent(getByText('Yucky Starbucks'), 'pressIn');
    fireEvent(getByText('Yucky Starbucks'), 'press');


    // fireEvent(getByText('Yucky Starbucks'), 'press');
    // expect(getByText('Yucky Starbucks')).not.toBeVisible();

  });

  it('Should call updateQuery when searchbar content changes', function (){

      const {getByTestId} = render(
          <GlobalContext.Provider value={globalContextMock}>
              <CustomSearchBar searchBarFocused={false} setSearchBarFocussed={jest.fn()}/>
          </GlobalContext.Provider>,
      );

      //I dont know why this throws an error about icons.
      // fireEvent.changeText(getByTestId('search-bar'), 'a');

  });
});