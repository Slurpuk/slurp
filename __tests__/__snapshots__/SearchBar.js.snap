// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search bar should render the search bar 1`] = `
<View
  style={
    Object {
      "alignSelf": "flex-end",
      "flex": 1,
      "marginRight": 0,
      "position": "relative",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "whitesmoke",
        "borderBottomColor": "#046D66",
        "borderBottomWidth": 2,
        "borderColor": "#046D66",
        "borderRadius": 20,
        "borderTopColor": "#046D66",
        "borderTopWidth": 2,
        "borderWidth": 2,
        "color": "#046D66",
        "height": 93.38000000000001,
        "marginLeft": "3%",
        "marginRight": "3%",
        "overflow": "visible",
        "padding": 8,
        "zIndex": 0,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "whitesmoke",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 20,
            "color": "#046D66",
            "flexDirection": "row",
            "height": 66.7,
            "minHeight": 30,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Where to..."
          placeholderTextColor="#046D66"
          platform="default"
          style={
            Object {
              "color": "#046D66",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
            }
          }
          testID="search-bar"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "color": "#046D66",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "marginLeft": -133.4,
        "maxHeight": 693.6800000000001,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "white",
          "height": 200.1,
          "opacity": 0.8,
          "position": "absolute",
          "top": -200.1,
          "width": 825.0000000000001,
          "zIndex": -1,
        }
      }
    />
    <RCTScrollView
      data={
        Array [
          Object {
            "IsOpen": true,
            "Name": "Yucky Starbucks",
          },
          Object {
            "IsOpen": true,
            "Name": "Black Goat",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      keyboardShouldPersistTaps="handled"
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      styles={
        Object {
          "display": "flex",
          "overflow": "visible",
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "width": 525,
                },
                Object {
                  "display": "flex",
                },
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                  },
                  Object {
                    "width": 825.0000000000001,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "black",
                    },
                    Object {
                      "borderBottomColor": "#DDDDDD",
                      "borderBottomWidth": 1,
                      "borderRadius": 0,
                      "fontFamily": "Poppins",
                      "fontSize": 16,
                      "height": 54,
                      "maxWidth": 750,
                      "paddingVertical": 15,
                      "position": "relative",
                      "textAlign": "center",
                      "textAlignVertical": "center",
                      "width": 750,
                    },
                  ]
                }
              >
                Yucky Starbucks
              </Text>
            </View>
          </View>
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "width": 525,
                },
                Object {
                  "display": "flex",
                },
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Array [
                  Object {
                    "backgroundColor": "white",
                  },
                  Object {
                    "width": 825.0000000000001,
                  },
                ]
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "black",
                    },
                    Object {
                      "borderBottomColor": "#DDDDDD",
                      "borderBottomWidth": 1,
                      "borderRadius": 0,
                      "fontFamily": "Poppins",
                      "fontSize": 16,
                      "height": 54,
                      "maxWidth": 750,
                      "paddingVertical": 15,
                      "position": "relative",
                      "textAlign": "center",
                      "textAlignVertical": "center",
                      "width": 750,
                    },
                  ]
                }
              >
                Black Goat
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
